CoW: add warning mode for cases that will change behaviour

jorisvandenbossche opened the pull request:
xref https://github.com/pandas-dev/pandas/issues/56019, https://github.com/pandas-dev/pandas/issues/48998

An initial PR working towards one of the remaining TODO items:

> Add a warning mode that gives deprecation warnings for all cases where the current behaviour would change (initially also behind an option) 

The current PR adds an option `pd.options.mode.copy_on_write = "warn"`, and then everything behaves as the current default (without CoW enabled) but will raise a warning in case the behaviour will change if CoW would be enabled.

As a first step, this PR adds the option, and adds the warning in one specific code path that can trigger CoW, i.e. updating a Series (`Manager.setitem_inplace`). So this covers a typical case like:

```
s = df["col"]
s[0] = ..
```

no longer updating `df`.

In addition, I also turned off SettingWithCopyWarning when the CoW warning mode is enabled (because in that case, those warnings are no longer correct for the future behaviour, and it seemed better to not raise both kinds of warnings at the same time)

jorisvandenbossche committed:
CoW: add warning mode for cases that will change behaviour

jorisvandenbossche commented:
There are still a whole bunch of failures (with the warning mode enabled, for which I still need to add a CI build), but quite some of them are due to inplace operators (eg `s += 1`) unnecessarily triggering CoW, which is something I will try to tackle as a precursor.

jorisvandenbossche committed:
add CI build

jorisvandenbossche committed:
avoid warning with inplace operators

phofl commented on pandas/core/frame.py:
```diff
@@ -4166,11 +4167,13 @@ def isetitem(self, loc, value) -> None:
                     _chained_assignment_msg, ChainedAssignmentError, stacklevel=2
                 )
+            elif warn_copy_on_write() and sys.getrefcount(self) <= 3:
+                warnings.warn("ChainedAssignmentError", FutureWarning, stacklevel=2)
```
Is the missing message intentional?

jorisvandenbossche replied:
No, just WIP, was focusing on the mechanics first

phofl commented on pandas/core/internals/managers.py:
```diff
@@ -1988,9 +1993,34 @@ def setitem_inplace(self, indexer, value) -> None:
-        if using_copy_on_write() and not self._has_no_reference(0):
-            self.blocks = (self._block.copy(),)
-            self._cache.clear()
+        if not self._has_no_reference(0):
```
I'd rather not check references before establishing that we are in the correct mode after our performance issues

jorisvandenbossche replied:
Good point, will update

phofl commented on pandas/core/internals/managers.py:
```diff
@@ -1988,9 +1993,34 @@ def setitem_inplace(self, indexer, value) -> None:
+                if cow_context == "chained-assignment":
+                    warnings.warn(
+                        "ChainedAssignmentError: behaviour will change in pandas 3.0 "
+                        "with Copy-on-Write ...",
```
I think we need a doc page that we can link to that explains this more in-depth

phofl commented on pandas/core/series.py:
```diff
@@ -1177,11 +1180,14 @@ def _get_value(self, label, takeable: bool = False):
                     _chained_assignment_msg, ChainedAssignmentError, stacklevel=2
                 )
+            elif warn_copy_on_write() and sys.getrefcount(self) <= 3:
+                cow_context = "chained-assignment"
```
I think you need to warn here instead of deferring. This might now warn, if the first op in chained assignment created a copy, e.g.

```
pd.options.mode.copy_on_write = "warn"

df = pd.DataFrame({"a": [1, 2, 3,], "b": 1})

indexer = df.a > 1
df = df.a
df[indexer][0] = 5
```

I think you *could* make a case for not warning here since the current behaviour is a no-op as well, but

- the DataFrame case warns
- I would prefer a warning since cow will also warn

jorisvandenbossche replied:
So the reason that I didn't yet warn here and deferred it to `Manager.setitem_inplace`, is because I wanted the warning to be more specific, so we can distinguish the case of:

```
df = ..
ser = df["col"]
ser[0] = 0
```
-> here we want to warn that you are setting into a series that is a view of another, and that this change will no longer propagate (in this case no longer update the dataframe)

```
df = ...
df["col"][0] = 0
```

-> here we want to warn that this simply won't work anymore in the future, as this is chained assignment (and the case of it that currently does work and is used widely I think, so I want to have a distinct warning for this one)

Now, while writing this I am not sure anymore why this alone warrants deferring it. I have to check in the tests for the case that made me move this to the lower level (I originally was just warning here, as you suggest, but then changed it)

phofl reviewed (commented):
first part of the review (wifi connection on the train isn't stable)

phofl commented on pandas/tests/copy_view/test_indexing.py:
```diff
@@ -159,10 +161,14 @@ def test_subset_column_slice(backend, using_copy_on_write, using_array_manager,
-        warn = SettingWithCopyWarning if single_block else None
+        # TODO
+        warn = (
+            SettingWithCopyWarning
```
I am confused, shouldn't this be silent when CoW is enabled as well?

jorisvandenbossche replied:
Yes, and that `using_copy_on_Write` case is already handled in the `if` block above, so in this `else` I still need to handle the `warn_copy_on_write` case. 

I could actually do a `elif warn_copy_on_write` separate block, that might be easier to read and which is also the pattern I started to use in many other places.

phofl commented on pandas/tests/copy_view/test_indexing.py:
```diff
@@ -318,7 +326,8 @@ def test_subset_set_with_row_indexer(backend, indexer_si, indexer, using_copy_on
         pytest.skip("setitem with labels selects on columns")
 
-    if using_copy_on_write:
+    # TODO
```
This should warn, shouldn't it?

phofl replied:
talking about the regular CoW mode

jorisvandenbossche replied:
No, for the regular CoW mode is shouldn't warn I think. The test is a bit hard to read, but it boils down to:

```
df = DataFrame({"a": [1, 2, 3, 4], "b": [4, 5, 6, 7], "c": [0.1, 0.2, 0.3, 0.4]})
subset = df[1:4]

subset.iloc[0:2] = 0
```

So this is a case that should warn in the warn mode (because it currently _does_ modify `df`), but it's not chained assignment, so the standard CoW rules apply and `df` doesn't get updated

phofl commented on pandas/tests/copy_view/test_indexing.py:
```diff
@@ -482,7 +493,8 @@ def test_subset_set_columns(backend, using_copy_on_write, dtype):
     subset = df[1:3]
 
-    if using_copy_on_write:
+    # TODO
```
Can you add an identifier to these TODOs, e.g. TODO(CoW-warn) or something like that.

phofl commented on pandas/tests/indexing/multiindex/test_setitem.py:
```diff
@@ -527,28 +532,34 @@ def test_frame_setitem_view_direct(
         with tm.raises_chained_assignment_error():
             df["foo"]["one"] = 2
+    elif warn_copy_on_write:
+        with tm.assert_produces_warning(FutureWarning, match="ChainedAssignmentError"):
```
Can we do something like ``tm.raises_chained_assignment_error``

phofl commented on pandas/tests/indexing/test_chaining_and_caching.py:
```diff
@@ -519,21 +563,21 @@ def test_detect_chained_assignment_warnings_errors(self, using_copy_on_write):
                 # INFO(CoW) no warning, and original dataframe not changed
-                with tm.assert_produces_warning(None):
-                    chained[2] = rhs
+                # with tm.assert_produces_warning(None):
```
?

jorisvandenbossche replied:
Sorry, left-over from testing

jorisvandenbossche committed:
update chained assignment tests to silence current failures

jorisvandenbossche committed:
address feedback

jorisvandenbossche commented:
@phofl I removed the last commit about the chained assignments again, and moved that to a separate PR (https://github.com/pandas-dev/pandas/pull/55522). So this PR then just focuses on adding the infrastructure, and the warnings for (non-chained) series setitem (the things going through `setitem_inplace`). Then once this is merged, I can work more easily in parallel on different areas.

jorisvandenbossche committed:
update warning message

jorisvandenbossche commented on pandas/core/internals/managers.py:
```diff
@@ -96,6 +99,20 @@
     from pandas.api.extensions import ExtensionArray
 
 
+COW_WARNING_SETITEM_MSG = """\
+Setting a value on a view: behaviour will change in pandas 3.0.
+Currently, the mutation will also have effect on the object that shares data
+with this object. For example, when setting a value in a Series that was
+extracted from a column of a DataFrame, that DataFrame will also be updated:
+
+    ser = df["col"]
+    ser[0] = 0     <--- in pandas 2, this also updates `df`
+
+In pandas 3.0 (with Copy-on-Write), updating one Series/DataFrame will never
+modify another, and thus in the example above, `df` will not be changed.
+"""
```
One question is: how elaborate do we want to make this warning message?

One the one hand, I think one can argue that it can be short and point to a doc page where it is explained in more detail (which I actually should add anyway). 

On the other hand, I also want it to be understandable for the many people that won't read the docs about this in detail.

mroeschke replied:
I would anticipate users also wanting to remove/address this warning, so in light of that, I would opt to move the explanation to a doc page if we can include potential actions users can take to remove this warning (e.g. set the copy-on-write mode to True)

jorisvandenbossche replied:
Is it OK to leave this as is, for now, and do a separate PR to write the documentation and update the warnings (add links to the docs, potentially shorten it)?

mroeschke replied:
Yeah sure thing 👍

phofl replied:
Agree with the doc page

jorisvandenbossche commented on pandas/tests/copy_view/test_indexing.py:
```diff
@@ -892,10 +906,14 @@ def test_column_as_series(backend, using_copy_on_write, using_array_manager):
-        warn = SettingWithCopyWarning if dtype_backend == "numpy" else None
-        with pd.option_context("chained_assignment", "warn"):
-            with tm.assert_produces_warning(warn):
+        if warn_copy_on_write:
+            with tm.assert_produces_warning(FutureWarning):
                 s[0] = 0
```
How important do we find it that we ensure the exact message? (given that this is only temporary, maybe not super important?)

(although I suppose it is not difficult to just add `match="Setting a value on a view"` to all cases I am adding in this PR.

I don't know how useful it will be to add a custom `tm.assert_produces_cow_warning`, but it would make the `tm.assert_produces_warning(FutureWarning, match="Setting a value on a view"):` a bit shorter. 

But when adding warnings for more cases, the exact warning message might also differ.

mroeschke replied:
I think it's okay being lax with matching the warning message during testing (but noting a comment that it's CoW related if it's not obvious from the test)

jorisvandenbossche replied:
> but noting a comment that it's CoW related if it's not obvious from the test

That might be a reason to actually make something like a `tm.assert_produces_cow_warning`, then it's also clear without adding a comment everywhere, and will be easier to remove all of those.

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/main' into cow-warnings

jorisvandenbossche committed:
fixup tests for changed warning text

jorisvandenbossche committed:
fix COW setitem

jorisvandenbossche committed:
Merge remote-tracking branch 'upstream/main' into cow-warnings

jorisvandenbossche committed:
add tm.assert_cow_warning and use everywhere

jorisvandenbossche commented:
I added a `tm.assert_cow_warning` helper function to use for all the cases where specifically checking for this warning.

@phofl @mroeschke if possible, I would like to get this merged shortly, as there is a lot more work to do here, but that will be easier done in parallel once this base PR is in.

phofl commented:
I’ll check later today or tomorrow

phofl commented:
thx @jorisvandenbossche
